src/lib.rs

```diff
@@ -203,6 +203,145 @@ pub struct Match<'t> {
     end: usize,
 }
 
+/// Returns the smallest possible index of the next valid UTF-8 sequence
+/// starting after `i`.
+/// Adapted from a function with the same name in the `regex` crate.
+fn next_utf8(text: &str, i: usize) -> usize {
+    let b = match text.as_bytes().get(i) {
+        None => return i + 1,
+        Some(&b) => b,
+    };
+    let inc = if b <= 0x7F {
+        1
+    } else if b <= 0b110_11111 {
+        2
+    } else if b <= 0b1110_1111 {
+        3
+    } else {
+        4
+    };
+    i + inc
```

Comment:
We already have a function called `codepoint_len` which we should be able to use here, to not duplicate this logic.

Reply:
I fixed this and moved `next_utf8` next to the other utility methods. Also added a test for zero length matches before longer codepoints (s. t. `next_utf8` is expected to increment by more than one).